src/ape/managers/chain.py

```diff
@@ -408,18 +414,82 @@ def outgoing(self) -> Iterator[ReceiptAPI]:
         """
         All outgoing transactions, from earliest to latest.
         """
-        explorer = self.provider.network.explorer
-        explorer_receipts = (
-            [r for r in explorer.get_account_transactions(self.address)] if explorer else []
-        )
-        for receipt in explorer_receipts:
-            if receipt.sender != self.address:
-                # Likely ``self.address`` is a contract.
-                # Cache the receipts by their sender instead and skip them here.
-                self.chain_manager.history.append(receipt)
-                continue
+
+        start_nonce = 0
+        stop_nonce = self.__len__() - 1  # just to cache this value
+
+        # TODO: Add ephemeral network sessional history to `ape-cache` instead,
+        #       and remove this (replace with `yield from iter(self[:len(self)])`)
+        for receipt in self.sessional:
+            if receipt.nonce < start_nonce:
+                raise QueryEngineError("Sessional history corrupted")
+
+            if receipt.nonce > start_nonce:
+                # NOTE: There's a gap in our sessional history, so fetch from query engine
+                yield from iter(self[start_nonce : receipt.nonce + 1])  # noqa: E203
 
             yield receipt
+            start_nonce = receipt.nonce + 1  # start next loop on the next item
+
+        if start_nonce != stop_nonce:
+            # NOTE: there is no more sessional history, so just return query engine iterator
+            yield from iter(self[start_nonce : stop_nonce + 1])  # noqa: E203
+
+    def query(
+        self,
+        *columns: List[str],
+        start_nonce: int = 0,
+        stop_nonce: Optional[int] = None,
+        engine_to_use: Optional[str] = None,
```

Comment:
When specifying a query engine that does not exist, it would nice if it told me some available ones, or maybe a closely matched one (using difflib and such).

Right now, it says:

```
QueryEngineError: Query engine `memory` not found.
```

I am not sure of any values to test here. I'd like to try to use the kwarg.

Reply:
`"__default__"` is the query one which will be handy for testing once the cache is enabled by default

Reply:
Added in [6e1501f](https://github.com/ApeWorX/ape/pull/1277/commits/6e1501f321c6552da8b27d67b8990efb1b69bfb1)